### PR TITLE
Ensure a switch update during allocation of multiple machines works

### LINKED ISSUE #193: Update of entities often produces "entity was changed from another"
This is a expected behaviour, but should not be visible to the API consumer.

examples:

```
{"level":"error","time":"2021-05-09T07:48:16.737Z","caller":"service/service.go:90","message":"service error","rqid":"7a5b83a95a9e25886b719bff8cf45a3d","remoteaddr":"10.20.2.6","method":"POST","uri":"/metal/v1/ip","route":"/metal/v1/ip/","useremail":"metal-admin@metal-stack.io","operation":"updateIP","status":422,"error":"cannot update ip (100.127.130.143): the entity was changed from another, please retry","service-caller":"ip-service.go:368","resp":"cannot update ip (100.127.130.143): the entity was changed from another, please retry (422)"}

```

```
{"level":"error","time":"2021-05-10T05:02:42.822Z","caller":"service/service.go:90","message":"service error","rqid":"8bad66d77ed59cc9634131c6b71affe4","remoteaddr":"10.20.4.7","method":"POST","uri":"/metal/v1/machine/register","route":"/metal/v1/machine/register","useremail":"metal-edit@metal-stack.io","operation":"registerMachine","status":422,"error":"cannot update switch (stg-kkw701-r01leaf02): the entity was changed from another, please retry","service-caller":"machine-service.go:657","resp":"cannot update switch (stg-kkw701-r01leaf02): the entity was changed from another, please retry (422)"}
```

### PR DESCRIPTION
**Pull request by majst01**
Call updateSwitch with a retry loop, closes #193

### CODE REVIEW

**Review by Gerrit91 — APPROVED** [review]

**Comment by Gerrit91** [comment]
I don't think this fully closes #193 though as this can still happen for machine free?

**Comment by majst01** [comment]
> I don't think this fully closes #193 though as this can still happen for machine free?

Cant reproduce the switch update failure even with a concurrency of 200 in https://github.com/metal-stack/metal-api/blob/master/cmd/metal-api/internal/service/machine-service_allocation_test.go#L150

**Comment by Gerrit91** [comment]
There is a pull request #129 with a concern that I think is still valid. Before closing #193, I think I would like to take a look at why certain switch operations are so expensive in production, why switch updates have a high risk of going wrong and if makes sense to also include this VRF delete into the retry mechanism.

**Comment by majst01** [comment]
Few additional Errors:

```
{"level":"error","time":"2022-03-07T15:00:53.275Z","caller":"service/machine-service.go:1912","message":"cannot delete vrf switches","machineID":"b70b4c00-0dd8-11eb-8000-3cecef6b3868","error":"cannot update switch (stg-kkw701-r02leaf01): the entity was changed from another, please retry"}
```

and

```
{"level":"error","time":"2022-03-07T15:00:53.275Z","caller":"service/service.go:96","message":"service error","rqid":"13878c1a137aae74e3058857a374be4a","remoteaddr":"10.20.9.5","method":"DELETE","uri":"/metal/v1/machine/b70b4c00-0dd8-11eb-8000-3cecef6b3868/free","route":"/metal/v1/machine/{id}/free","useremail":"metal-admin@metal-stack.io","operation":"freeMachine","status":422,"error":"cannot delete vrf switches: cannot update switch (stg-kkw701-r02leaf01): the entity was changed from another, please retry","service-caller":"machine-service.go:1693","resp":"cannot delete vrf switches: cannot update switch (stg-kkw701-r02leaf01): the entity was changed from another, please retry (422)"}

```

and

```
{"level":"error","time":"2022-03-07T17:24:16.616Z","caller":"service/service.go:96","message":"service error","rqid":"f0c63346c467c415002bf5116b2608a3","remoteaddr":"10.20.11.3","method":"POST","uri":"/metal/v1/switch/stg-kkw701-r01leaf02/notify","route":"/metal/v1/switch/{id}/notify","useremail":"metal-edit@metal-stack.io","operation":"notifySwitch","status":422,"error":"cannot update switch (stg-kkw701-r01leaf02): the entity was changed from another, please retry","service-caller":"switch-service.go:180","resp":"cannot update switch (stg-kkw701-r01leaf02): the entity was changed from another, please retry (422)"}
```

**Comment by majst01** [comment]
Should cover now all important switch related errors seen in the metal-api hurting machine livecycle

File: cmd/metal-api/internal/service/switch-service.go [hunk 1]
```diff
@@ -176,6 +177,7 @@ func (r switchResource) notifySwitch(request *restful.Request, response *restful
 		s.LastSyncError = sync
 	}
 
+	// FIXME needs retry coverage, not so urgent, will be sent in 5 minutes again
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/switch-service.go:180]: This part could be "easily" mitigated by writing the switch status into a separate table.

Comment (Gerrit91) [inline cmd/metal-api/internal/service/switch-service.go:180]: This part could be "easily" mitigated by writing the switch status into a separate table.

File: cmd/metal-api/internal/service/machine-service.go [hunk 2]
```diff
@@ -1753,6 +1774,7 @@ func (r machineResource) deleteMachine(request *restful.Request, response *restf
 		}
 		delete(newIP.MachineConnections, m.ID)
 
+		// FIXME needs retry coverage, not so urgent, deleteMachine is called from metalctl, error is promoted and can be retried
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go:1777]: This is a pure operator endpoint, always manual and only on dead machines... no need for retry from my point of view.

File: cmd/metal-api/internal/service/switch-service.go (hunk 1 above)
Comment (Gerrit91) [reply cmd/metal-api/internal/service/switch-service.go:180]: https://github.com/metal-stack/metal-api/issues/263

**Review by Gerrit91 — APPROVED** [review]

File: cmd/metal-api/internal/service/switch-service.go [hunk 3]
```diff
@@ -231,7 +235,18 @@ func (r switchResource) updateSwitch(request *restful.Request, response *restful
 
 	newSwitch.Mode = metal.SwitchModeFrom(requestPayload.Mode)
 
-	err = r.ds.UpdateSwitch(oldSwitch, &newSwitch)
+	err = retry.Do(
+		func() error {
+			err := r.ds.UpdateSwitch(oldSwitch, &newSwitch)
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service.go:240]: This is wrong, switch must be fetched again

File: cmd/metal-api/internal/service/switch-service.go [hunk 4]
```diff
@@ -323,7 +338,18 @@ func (r switchResource) registerSwitch(request *restful.Request, response *restf
 		s.Nics = nics
 		// Do not replace connections here: We do not want to loose them!
 
-		err = r.ds.UpdateSwitch(&old, s)
+		err = retry.Do(
+			func() error {
+				err := r.ds.UpdateSwitch(&old, s)
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service.go:343]: This is wrong, switch must be fetched again

**Pull request merged**